js/stores/message_store.js

```diff
@@ -39,7 +56,26 @@ DispatchHandler[ActionTypes.EDIT_LAST_MESSAGE] = function(data) {
 };
 
 DispatchHandler[ActionTypes.NEW_MESSAGE] = function(message) {
-  _messages.push(message);
+  messageCopy = JSON.parse(JSON.stringify(message)); // Cloning the object
+  messageCopy.status = "Pending";
+  messageCopy.key = messageCopy.clientId;
+  _pendingMessages[message.clientId] = messageCopy;
+};
+
+DispatchHandler[ActionTypes.CREATE_MESSAGE_SUCCESS] = function(message) {
+  delete _pendingMessages[message.clientId];
+  message.status = "Success";
+  _successfulMessages.push(message);
+};
+
+DispatchHandler[ActionTypes.INCOMING_MESSAGE] = function(message) {
+  if (!_pendingMessages.hasOwnProperty(message.clientId)) { //Filter out local client messages
+    _successfulMessages.push(message);
+  }
+};
+
+DispatchHandler[ActionTypes.CREATE_MESSAGE_ERROR] = function(data) {
+  _pendingMessages[message.clientId].status = "Failed";
 };
```

Comment:
flash "Your message failed to send." (and maybe delete from pending after a couple seconds?)